pkg/network/prober/prober.go

```diff
@@ -26,40 +26,70 @@ import (
 	"github.com/pkg/errors"
 	"k8s.io/apimachinery/pkg/util/sets"
 	"k8s.io/apimachinery/pkg/util/wait"
-
-	"github.com/knative/serving/pkg/network"
 )
 
 // TransportFactory is a function which returns an HTTP transport.
 type TransportFactory func() http.RoundTripper
 
-// ProbeOption is a way for caller to modify the HTTP request before it goes out.
-type ProbeOption func(r *http.Request) *http.Request
+// Option provides a way for the caller to mutate the HTTP request before it goes out,
+// and to validate the HTTP response after it comes back.
+type Option struct {
+	Prepare func(r *http.Request) *http.Request
+	Validate func(r *http.Response) (bool, error)
+}
+
+// WithHeader sets a header in the probe request.
+func WithHeader(name, value string) Option {
+	return Option{
+		Prepare: func(r *http.Request) *http.Request {
+			r.Header.Set(name, value)
+			return r
+		},
+	}
+}
+
+// ExpectsBody validates that the body of the probe response matches the provided string.
+func ExpectsBody(body string) Option {
+	return Option{
+		Validate: func(r *http.Response) (bool, error) {
+			bytes, err := ioutil.ReadAll(r.Body)
```

Comment:
This is going to be a problem, if there are many validators checking body.
Probably body validators should accept `[]byte`.

Reply:
But then it has to become a `func(r *http.Response, b []byte) (bool, error)`.
What is the scenario for multiple Body checks?
I can change it, it is just less clean.

Reply:
I don't know, but if you're opening that window of possibility, you have to be correct.
Unless you split body check into different method... :-)

Reply:
Changed.